DeepSpeech.py

```diff
@@ -694,594 +651,999 @@ def stopwatch(start_duration=0):
+        saver.restore(session, checkpoint_path)
+        log_info('Restored checkpoint at training epoch %d' % (int(checkpoint_path.split('-')[-1]) + 1))
+
+        # Initialise the model exporter and export the model
+        model_exporter.init(session.graph.as_graph_def(),
+                            named_graph_signatures = {
+                                'inputs': exporter.generic_signature(
+                                    { 'input': input_tensor }),
+                                'outputs': exporter.generic_signature(
+                                    { 'outputs': decoded})})
+        if FLAGS.remove_export:
+            actual_export_dir = os.path.join(FLAGS.export_dir, '%08d' % FLAGS.export_version)
+            if os.path.isdir(actual_export_dir):
+                log_info('Removing old export')
+                shutil.rmtree(actual_FLAGS.export_dir)
+        try:
+            # Export serving model
+            model_exporter.export(FLAGS.export_dir, tf.constant(FLAGS.export_version), session)
+
+            # Export graph
+            input_graph_name = 'input_graph.pb'
+            tf.train.write_graph(session.graph, FLAGS.export_dir, input_graph_name, as_text=False)
+
+            # Freeze graph
+            input_graph_path = os.path.join(FLAGS.export_dir, input_graph_name)
+            input_saver_def_path = ''
+            input_binary = True
+            output_node_names = 'output_node'
+            restore_op_name = 'save/restore_all'
+            filename_tensor_name = 'save/Const:0'
+            output_graph_path = os.path.join(FLAGS.export_dir, 'output_graph.pb')
+            clear_devices = False
+            freeze_graph.freeze_graph(input_graph_path, input_saver_def_path,
+                                      input_binary, checkpoint_path, output_node_names,
+                                      restore_op_name, filename_tensor_name,
+                                      output_graph_path, clear_devices, '')
+
+            log_info('Models exported at %s' % (FLAGS.export_dir))
+        except RuntimeError:
+            log_error(sys.exc_info()[1])
+
+
+def main(_) :
+
+    # Determine, if we are the chief worker
+    global is_chief
+    is_chief = len(FLAGS.worker_hosts) == 0 or (FLAGS.task_index == 0 and FLAGS.job_name == 'worker')
+
+    global COORD
+    COORD = TrainingCoordinator()
+    COORD.start()
+
+    # Set default relative data set directory based on importer name
+    if FLAGS.dataset_path == '':
+        FLAGS.dataset_path = os.path.join('data', FLAGS.importer)
+
+    # Lazy-import data set module
+    global importer_module
+    importer_module = importlib.import_module('util.importers.%s' % FLAGS.importer)
+
+    from util.website import maybe_publish
+    if FLAGS.fulltrace:
+        check_cupti()
+
+    # ps and worker hosts required for p2p cluster setup
+    FLAGS.ps_hosts = filter(len, FLAGS.ps_hosts.split(','))
+    FLAGS.worker_hosts = filter(len, FLAGS.worker_hosts.split(','))
```

Comment:
For Python 2/3 compat, these need to be `list(filter...`, and then we need to import `six.moves.filter`.